### PR TITLE
CrossValidationSplit in mlContext.Data

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Just a QoL change.
fixes https://github.com/dotnet/machinelearning/issues/3049

### CODE REVIEW

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 1]
```diff
@@ -472,6 +525,18 @@ internal static void EnsureGroupPreservationColumn(IHostEnvironment env, ref IDa
                         columnOptions = new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30);
                     data = new HashingEstimator(env, columnOptions).Fit(data).Transform(data);
                 }
+                else
+                {
+                    if (!data.Schema[samplingKeyColumn].IsNormalized() && type == NumberDataViewType.Single || type == NumberDataViewType.Double)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >type == NumberDataViewType.Single || type == NumberDataViewType.Double [](start = 74, length = 70)

brackets around

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 2]
```diff
@@ -433,6 +433,60 @@ public TrainTestData TrainTestSplit(IDataView data, double testFraction = 0.1, s
...
+        /// If <see langword="null"/> no row grouping will be performed.</param>
+        /// <param name="seed">Seed for the random number generator used to select rows for cross-validation folds.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[TrainTestSplit](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public List<TrainTestData> CrossValidationSplit(IDataView data, int numberOfFolds = 5, string samplingKeyColumnName = null, int? seed = null)
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >List [](start = 15, length = 4)

`IEnumerable` or `IList` so we have less to hold onto?

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 3]
```diff
@@ -433,6 +433,60 @@ public TrainTestData TrainTestSplit(IDataView data, double testFraction = 0.1, s
...
+            _env.CheckParam(numberOfFolds > 1, nameof(numberOfFolds), "Must be more than 1");
+            _env.CheckValueOrNull(samplingKeyColumnName);
+            EnsureGroupPreservationColumn(_env, ref data, ref samplingKeyColumnName, seed);
+            var result = new List<TrainTestData>();
+            foreach (var split in CrossValidationSplit(_env, data, numberOfFolds, samplingKeyColumnName))
+                result.Add(split);
+            return result;
+        }
+
+        internal static IEnumerable<TrainTestData> CrossValidationSplit(IHostEnvironment env, IDataView data, int numberOfFolds = 5, string samplingKeyColumnName = null)
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >internal static IEnumerable<TrainTestData> [](start = 8, length = 42)

Just a note: Eventually, I think we should move these operations out of the catalog and into a core library.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 4]
```diff
@@ -472,6 +525,18 @@ internal static void EnsureGroupPreservationColumn(IHostEnvironment env, ref IDa
...
                     data = new HashingEstimator(env, columnOptions).Fit(data).Transform(data);
                 }
+                else
+                {
+                    if (!data.Schema[samplingKeyColumn].IsNormalized() && type == NumberDataViewType.Single || type == NumberDataViewType.Double)
+                    {
+                        var origStratCol = samplingKeyColumn;
+                        int inc = 0;
+                        while (data.Schema.TryGetColumnIndex(samplingKeyColumn, out int tmp))
+                            samplingKeyColumn = string.Format("{0}_{1:000}", origStratCol, ++inc);
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >000 [](start = 70, length = 3)

This specifies 1000 tries, no?

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: I have better idea!

---
In reply to: [267986740](https://github.com/dotnet/machinelearning/pull/3058#discussion_r267986740) [](ancestors = 267986740)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            // Creating the ML.Net IHostEnvironment object, needed for the pipeline.
+            var mlContext = new MLContext();
+
+            // Generate some data points.
+            var examples = GenerateRandomDataPoints(10);
+
+            // Convert the examples list to an IDataView object, which is consumable by ML.NET API.
+            var dataview = mlContext.Data.LoadFromEnumerable(examples);
+            // Cross validation splits your data randomly into set of chunks, and set test set in each fold to be certain chunk of data
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: Maybe

" // Cross validation splits your data randomly into set of chunks, and set test set in each fold to be certain chunk of data and training data set to be rest of the data."

to "Cross validation splits your data randomly into set of "folds", and creates groups of Train and Test sets, where for each group, one fold is the Test and the rest of the folds the Train."

"

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 6]
```diff
@@ -433,6 +433,60 @@ public TrainTestData TrainTestSplit(IDataView data, double testFraction = 0.1, s
...
+        /// <param name="data">The dataset to split.</param>
+        /// <param name="numberOfFolds">Number of cross-validation folds.</param>
+        /// <param name="samplingKeyColumnName">Name of a column to use for grouping rows. If two examples share the same value of the <paramref name="samplingKeyColumnName"/>,
+        /// they are guaranteed to appear in the same subset (train or test). This can be used to ensure no label leakage from the train to the test set.
+        /// If <see langword="null"/> no row grouping will be performed.</param>
+        /// <param name="seed">Seed for the random number generator used to select rows for cross-validation folds.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[TrainTestSplit](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs)]
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: `CrossValidationSplit` instead of `TrainTestSplit` in the name.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            // Creating the ML.Net IHostEnvironment object, needed for the pipeline.
+            var mlContext = new MLContext();
+
+            // Generate some data points.
+            var examples = GenerateRandomDataPoints(10);
+
+            // Convert the examples list to an IDataView object, which is consumable by ML.NET API.
+            var dataview = mlContext.Data.LoadFromEnumerable(examples);
+            // Cross validation splits your data randomly into set of "folds", and creates groups of Train and Test sets,
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: newline

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            // [Group, 2], [Features, 0.5588848]
+            // [Group, 0], [Features, 0.9060271]
+            // [Group, 1], [Features, 0.4421779]
+            // [Group, 2], [Features, 0.9775497]
+            // [Group, 0], [Features, 0.2737045]
+            // 
+            // The data in the Test split.
+            // [Group, 2], [Features, 0.7680227]
+            // [Group, 2], [Features, 0.5588848]
+            // [Group, 2], [Features, 0.9775497]
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: is it correct? It shows trainset has 10 example already???

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            // [Group, 2], [Features, 0.5588848]
+            // [Group, 1], [Features, 0.4421779]
+            // [Group, 2], [Features, 0.9775497]
+            // 
+            // The data in the Test split.
+            // [Group, 0], [Features, 0.7262433]
+            // [Group, 0], [Features, 0.5581612]
+            // [Group, 0], [Features, 0.9060271]
+            // [Group, 0], [Features, 0.2737045]
+            PrintPreviewRows(folds[1]);
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: newline???

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                {
+                    Group = i % 3,
+
+                    // Create random features that are correlated with label.
+                    Features = (float)random.NextDouble()
+                };
+            }
+        }
+
+        // Example with label and group column. A data set is a collection of such examples.
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: >Example with label and group column. [](start = 11, length = 36)

No label. Only group and features.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs (hunk 8 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: Good catch, let me debug this.

---
In reply to: [268364243](https://github.com/dotnet/machinelearning/pull/3058#discussion_r268364243) [](ancestors = 268364243)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            // [Group, 2], [Features, 0.7680227]
+            // [Group, 0], [Features, 0.5581612]
+            // [Group, 1], [Features, 0.2060332]
+            // [Group, 1], [Features, 0.4421779]
+            // [Group, 2], [Features, 0.9775497]
+            // [Group, 0], [Features, 0.2737045]
+            // 
+            // The data in the Test split.
+            // [Group, 2], [Features, 0.5588848]
+            // [Group, 0], [Features, 0.9060271]
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: This also does not look good. Can you check this as well?

Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: Why?
Data got broken into 4/4/2 chunks.
You not happy with distribution?

---
In reply to: [268366691](https://github.com/dotnet/machinelearning/pull/3058#discussion_r268366691) [](ancestors = 268366691)

Comment (zeahmed) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/CrossValidationSplit.cs]: OK, that's TrainTestSplit in each fold are not expected to be of same size right?

---
In reply to: [268367175](https://github.com/dotnet/machinelearning/pull/3058#discussion_r268367175) [](ancestors = 268367175,268366691)

**Comment by daholste** [comment]
@Ivanidzo4ka Wow!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@justinormont @vinodshanbhag 
@Ivanidzo4ka -- we cannot thank you enough for this. Thanks for being a boss

**Comment by daholste** [comment]
@Ivanidzo4ka -- you completed this ONE DAY after a quick ping about if this exists. Can't thank you enough. Complete boss

**Comment by daholste** [comment]
You saved us from a recursive / tree-based crazy method of repeated Train/Test splits. Thank you so much

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 12]
```diff
@@ -459,19 +516,24 @@ internal static void EnsureGroupPreservationColumn(IHostEnvironment env, ref IDa
...
                     HashingEstimator.ColumnOptions columnOptions;
                     if (seed.HasValue)
                         columnOptions = new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30, (uint)seed.Value);
                     else
                         columnOptions = new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30);
                     data = new HashingEstimator(env, columnOptions).Fit(data).Transform(data);
                 }
+                else
+                {
+                    if (!data.Schema[samplingKeyColumn].IsNormalized() && (type == NumberDataViewType.Single || type == NumberDataViewType.Double))
```
Comment (daholste) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: @Ivanidzo4ka -- and on top of all of this, you spontaneously fixed a potential hole. amazing
